pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/nix.py

```diff
@@ -1,42 +1,168 @@
+import logging
 import os
 from datetime import datetime
+from importlib.resources import files
 from pathlib import Path
+from string import Template
 from subprocess import PIPE, CalledProcessError
 from typing import Final
 
 from .models import (
     Action,
+    BuildAttr,
     Flake,
     Generation,
     GenerationJson,
     NRError,
     Profile,
     Remote,
 )
-from .process import run_wrapper
-from .utils import Args, dict_to_flags, info
+from .process import SSH_DEFAULT_OPTS, run_wrapper
+from .utils import Args, dict_to_flags
 
 FLAKE_FLAGS: Final = ["--extra-experimental-features", "nix-command flakes"]
+FLAKE_REPL_TEMPLATE: Final = "repl.nix.template"
+logger = logging.getLogger(__name__)
+
+
+def build(
+    attr: str,
+    build_attr: BuildAttr,
+    **build_flags: Args,
+) -> Path:
+    """Build NixOS attribute using classic Nix.
+
+    Returns the built attribute as path.
+    """
+    run_args = [
+        "nix-build",
+        build_attr.path,
+        "--attr",
+        build_attr.to_attr(attr),
+        *dict_to_flags(build_flags),
+    ]
+    r = run_wrapper(run_args, stdout=PIPE)
+    return Path(r.stdout.strip())
+
+
+def build_flake(
+    attr: str,
+    flake: Flake,
+    **flake_build_flags: Args,
+) -> Path:
+    """Build NixOS attribute using Flakes.
+
+    Returns the built attribute as path.
+    """
+    run_args = [
+        "nix",
+        *FLAKE_FLAGS,
+        "build",
+        "--print-out-paths",
+        flake.to_attr(attr),
+        *dict_to_flags(flake_build_flags),
+    ]
+    r = run_wrapper(run_args, stdout=PIPE)
+    return Path(r.stdout.strip())
+
+
+def remote_build(
+    attr: str,
+    build_attr: BuildAttr,
+    build_host: Remote | None,
+    build_flags: dict[str, Args] | None = None,
+    instantiate_flags: dict[str, Args] | None = None,
+    copy_flags: dict[str, Args] | None = None,
+) -> Path:
+    r = run_wrapper(
+        [
+            "nix-instantiate",
+            "--raw",
```

Comment:
Is that a new arg to `nix-instantiate`?

On `nixos-24.11` with nix `2.24.10` I get the following error:
```
$ nixos-rebuild-ng build -f ../default.nix -A bob --target-host ssh://bob --build-host ssh://bob -vv
building the system configuration...
debug: nixos_rebuild.process: calling run with args=['nix-instantiate', '--raw', PosixPath('../default.nix'), '--attr', 'bob.config.system.build.toplevel', '-vv'], kwargs={'stdout': -1}, extra_env=None
error: unrecognised flag '--raw'
Try 'nix-instantiate --help' for more information.
```

Reply:
Almost sure that this was a brain fart, sorry. I think `--raw` is only for the new CLI.

I will take a look once I get home and should push a fix in #361750.

Reply:
I continued the discussion on the other PR. Thank you! :-)